// @wagmi/core/dist/index.js
import "./chunk-KX4UEHS5.js";
import {
  ChainMismatchError,
  ChainNotConfiguredError,
  Config,
  ConfigChainsNotFound,
  ConnectorAlreadyConnectedError,
  ConnectorNotFoundError,
  SwitchChainNotSupportedError,
  configureChains,
  connect,
  createConfig,
  createStorage,
  deepEqual,
  deserialize,
  disconnect,
  erc20ABI,
  erc4626ABI,
  erc721ABI,
  fetchBalance,
  fetchBlockNumber,
  fetchEnsAddress,
  fetchEnsAvatar,
  fetchEnsName,
  fetchEnsResolver,
  fetchFeeData,
  fetchToken,
  fetchTransaction,
  getAccount,
  getConfig,
  getContract,
  getNetwork,
  getPublicClient,
  getUnit,
  getWalletClient,
  getWebSocketPublicClient,
  multicall,
  noopStorage,
  prepareSendTransaction,
  prepareWriteContract,
  readContract,
  readContracts,
  sendTransaction,
  serialize,
  signMessage,
  signTypedData,
  switchNetwork,
  waitForTransaction,
  watchAccount,
  watchBlockNumber,
  watchContractEvent,
  watchMulticall,
  watchNetwork,
  watchPendingTransactions,
  watchPublicClient,
  watchReadContract,
  watchReadContracts,
  watchWalletClient,
  watchWebSocketPublicClient,
  writeContract
} from "./chunk-OKULXQAF.js";
import {
  mainnet,
  sepolia
} from "./chunk-JNJVMJJ2.js";
import {
  Connector
} from "./chunk-BVC4KGLQ.js";
import {
  InjectedConnector
} from "./chunk-EQOEZP46.js";
import "./chunk-MQXBDTVK.js";
export {
  ChainMismatchError,
  ChainNotConfiguredError,
  Config,
  ConfigChainsNotFound,
  Connector,
  ConnectorAlreadyConnectedError,
  ConnectorNotFoundError,
  InjectedConnector,
  SwitchChainNotSupportedError,
  configureChains,
  connect,
  createConfig,
  createStorage,
  deepEqual,
  deserialize,
  disconnect,
  erc20ABI,
  erc4626ABI,
  erc721ABI,
  fetchBalance,
  fetchBlockNumber,
  fetchEnsAddress,
  fetchEnsAvatar,
  fetchEnsName,
  fetchEnsResolver,
  fetchFeeData,
  fetchToken,
  fetchTransaction,
  getAccount,
  getConfig,
  getContract,
  getNetwork,
  getPublicClient,
  getUnit,
  getWalletClient,
  getWebSocketPublicClient,
  mainnet,
  multicall,
  noopStorage,
  prepareSendTransaction,
  prepareWriteContract,
  readContract,
  readContracts,
  sendTransaction,
  sepolia,
  serialize,
  signMessage,
  signTypedData,
  switchNetwork,
  waitForTransaction,
  watchAccount,
  watchBlockNumber,
  watchContractEvent,
  watchMulticall,
  watchNetwork,
  watchPendingTransactions,
  watchPublicClient,
  watchReadContract,
  watchReadContracts,
  watchWalletClient,
  watchWebSocketPublicClient,
  writeContract
};
