// viem/dist/cjs/actions/public/getFilterChanges.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getFilterChanges = void 0;
const decodeEventLog_js_1 = require("../../utils/abi/decodeEventLog.js");
const log_js_1 = require("../../utils/formatters/log.js");
async function getFilterChanges(_client, { filter }) {
    const logs = await filter.request({
        method: 'eth_getFilterChanges',
        params: [filter.id],
    });
    return logs
        .map((log) => {
        if (typeof log === 'string')
            return log;
        try {
            const { eventName, args } = 'abi' in filter && filter.abi
                ? (0, decodeEventLog_js_1.decodeEventLog)({
                    abi: filter.abi,
                    data: log.data,
                    topics: log.topics,
                })
                : { eventName: undefined, args: undefined };
            return (0, log_js_1.formatLog)(log, { args, eventName });
        }
        catch {
            return;
        }
    })
        .filter(Boolean);
}
exports.getFilterChanges = getFilterChanges;
